cmd/snap/cmd_run.go

```diff
@@ -748,7 +752,72 @@ func activateXdgDocumentPortal(info *snap.Info, snapApp, hook string) error {
 
 type envForExecFunc func(extra map[string]string) []string
 
+var gdbServerWelcomeFmt = `
+Welcome to "snap run --gdbserver".
+You are right before your application is run.
+Please open a different terminal and run:
+
+gdb -ex="target remote %[1]s" -ex=continue -ex="signal SIGCONT" -ex=continue
+(gdb) continue
+
+or use your favorite gdb frontend and connect to %[1]s
+`
+
+func racyFindFreePort() (int, error) {
+	l, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return 0, err
+	}
+	defer l.Close()
+	return l.Addr().(*net.TCPAddr).Port, nil
+}
+
+func (x *cmdRun) useGdbserver() bool {
+	return x.ParserRan == 1 && x.Gdbserver != "no-gdbserver"
```

Comment:
This looks a bit magical, why do we need to check if ParserRan and not query x.Gdbserver directly below, can you add a comment?

Reply:
This makes sure that go-flag parser actually ran and assigned values to options. Otherwise we'd get `Gdbserver=""` what could mean that someone passed `--gdbserver` even if they did not.

Reply:
I'll add a comment.